### PR TITLE
Add packed sequences for HFDatasetDataModule

### PR DESCRIPTION
**Pull request by athitten**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds support for sequence packing in HFDatasetDataModule class for automodel.
1) Creates a method `pack()` in `HFDatasetDataModule` class that calls the `pack` method from `HFDatasetPackedSequence` class which implements the packing logic and contains all the util functions used by `pack()` method.
2) Modifies `collate_fn` of `HFDatasetDataModule` class to also get the `attention_mask` from the packed batch which is the arg of the `collate_fn`.
3) Creates its own `collate_fn` for HFMockDataModule class without the packing related code.
 
**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/data/hf_dataset.py [hunk 1]
```diff
@@ -297,19 +303,29 @@ def __init__(
...
+        self.packed_sequence_size = packed_sequence_size
 
     @staticmethod
     def from_dict(dataset_dict, split, **kwargs):
         """wraps Dataset's from_dict method"""
         dataset = Dataset.from_dict(dataset_dict)
         return HFDatasetDataModule(path_or_dataset=dataset, split=split, **kwargs)
 
-    @staticmethod
-    def collate_fn(batch, pad_token_id=0, pad_seq_len_divisible=None):
```
Comment (athitten) [inline nemo/collections/llm/gpt/data/hf_dataset.py:311]: Make `collate_fn` a class method (instead of static) to be able to access instance variables like `packed_sequence_size` inside the `collate_fn`.

File: nemo/collections/llm/gpt/data/hf_dataset.py [hunk 2]
```diff
@@ -284,7 +290,7 @@ def __init__(
         self.dataset_splits = make_dataset_splits(dataset, split, split_aliases)
 
         if collate_fn is None:
-            self._collate_fn = lambda x: HFDatasetDataModule.collate_fn(
+            self._collate_fn = lambda x: self.collate_fn(
```
Comment (athitten) [inline nemo/collections/llm/gpt/data/hf_dataset.py:293]: `self.collate_fn` since `collate_fn` is made an instance method and not a static method.

File: nemo/collections/llm/gpt/data/hf_dataset.py [hunk 3]
```diff
@@ -574,7 +605,7 @@ def __init__(
         self.pin_memory = pin_memory
         self.persistent_workers = persistent_workers
         self.create_attention_mask = create_attention_mask
-        self.collate_fn = lambda x: HFDatasetDataModule.collate_fn(x, pad_token_id=0)
+        self.collate_fn = lambda x: HFMockDataModule.collate_fn(x, pad_token_id=0)
```
Comment (athitten) [inline nemo/collections/llm/gpt/data/hf_dataset.py:608]: Create an explicit and simple `collate_fn` for `HFMockDataModule` as the one from `HFDatasetDataModule` is not static anymore

**Review by BoxiangW — APPROVED** [review]

File: examples/llm/sft/automodel.py [hunk 4]
```diff
@@ -67,12 +69,17 @@ def formatting_prompts_func(example):
...
         pad_seq_len_divisible=16 if fp8 else None,  # FP8 training requires seq length to be divisible by 16.
+        packed_sequence_size=packed_sequence_size,
     )
+    ## tokenization is happening here
     datamodule.map(
         formatting_prompts_func,
         batched=False,
         remove_columns=["id", "title", "context", "question", 'answers'],
     )
+    # Pack the sequences in the dataset if packed_sequence_size > 0
```
Comment (akoumpa) [inline examples/llm/sft/automodel.py:80]: I'm wondering if there's a way to hide the `.pack` call within the `HFDatasetDataModule` class? For example, by moving this check just before the creation of the dataloaders (therefore all the .map calls have been applied) or similarly? The scenario I have in mind is a user wants to use packed sequences and they set the `packed_sequence_size` parameter accordingly but for some reason they forgot to call `.pack`. I'm still reading the code so I may have missed something. Please let me know what you think @athitten .

File: nemo/collections/llm/gpt/data/hf_dataset_packed_sequence.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+import torch
+from datasets import Dataset
+from torch.nn import functional as F
+from tqdm import tqdm
+
+CROSS_ENTROPY_IGNORE_IDX = -100
+PACK_TYPE = Dict[str, Union[torch.Tensor, List[int]]]
+
+
+class HFDatasetPackedSequence:
```
Comment (akoumpa) [inline nemo/collections/llm/gpt/data/hf_dataset_packed_sequence.py:26]: I'm wondering if it would be better to subclass HFDataset here instead, WDYT @athitten ?

**Comment by akoumpa** [comment]
Overall this looks great and implements an important feature, congrats @athitten .

I made a few minor comments, apart from the those, if there's time it would be great to include some unit tests (e.g. via `pytest`), to ensure that future changes do not break any code.

File: examples/llm/sft/automodel.py (hunk 4 above)
Comment (athitten) [reply examples/llm/sft/automodel.py:80]: Thanks for the review @akoumpa ! This a good point. Thinking of it, it should be possible to call packing method [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/gpt/data/hf_dataset.py#L342) if `packed_sequence_size > 0`. Let me try to make this change and check.

File: nemo/collections/llm/gpt/data/hf_dataset_packed_sequence.py (hunk 5 above)
Comment (athitten) [reply nemo/collections/llm/gpt/data/hf_dataset_packed_sequence.py:26]: The intention of `HFDatasetPackedSequence` is just to have instance variables so that they can be accessed within all the util methods of the main pack method like `_should_stop_packing`, `_split_and_add_pack` etc., Having each of these methods as separate functions outside of a class can make the function definitions very long leading to a not very clean code. With this being the only intention, having `HFDatasetPackedSequence` subclass `HFDatasetDataModule` won't be of any help. Also the main pack method is already part of `HFDatasetDataModule` class [here](https://github.com/NVIDIA/NeMo/blob/athitten/packed_sequence_automodel/nemo/collections/llm/gpt/data/hf_dataset.py#L418). So `HFDatasetPackedSequence` is more of a helper class here. Let me know if this is okay. Maybe I can mention it as `helper` in the class name so that its not confusing ?

**Comment by athitten** [comment]
> Overall this looks great and implements an important feature, congrats @athitten .
> 
> I made a few minor comments, apart from those, if there's time it would be great to include some unit tests (e.g. via `pytest`), to ensure that future changes do not break any code.

Yes unit tests and docs were going to be added in a subsequent PR soon.

File: examples/llm/sft/automodel.py (hunk 4 above)
Comment (athitten) [reply examples/llm/sft/automodel.py]: Added packing in [_make_dataloader](https://github.com/NVIDIA/NeMo/pull/12551/files#diff-ec2cdd55745a4a56dff9696f648b829d311ee8d7676a73d761adc0764153f011R447) method of `HFDatasetDataModulePacked` class.

File: nemo/collections/llm/gpt/data/hf_dataset_packed_sequence.py (hunk 5 above)
Comment (athitten) [reply nemo/collections/llm/gpt/data/hf_dataset_packed_sequence.py]: I did end up creating a [HFDatasetDataModulePacked](https://github.com/NVIDIA/NeMo/pull/12551/files#diff-ec2cdd55745a4a56dff9696f648b829d311ee8d7676a73d761adc0764153f011R411) class that overrides certain methods from HFDatasetDataModule class to support packing. This class gets instantiated in `examples/llm/sft/automodel.py`  if packed_sequence_size > 0

**Review by akoumpa — DISMISSED** [review]
LGTM, thank you.

**Review by akoumpa — APPROVED** [review]

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 6]
```diff
@@ -291,9 +291,16 @@ def training_step(self, batch, batch_idx=None, context_parallel=False):
             batch['input_ids'] = batch['tokens']
 
         # TODO(@boxiangw): Refractor. Needed for SP support
-        batch["position_ids"] = torch.arange(0, batch['input_ids'].shape[1]).unsqueeze(0).to(self.model.device)
+        # We cant have this for packed seq, attn_mask not in batch indicates regular sequence (not packed)
+        if 'attention_mask' not in batch:
+            batch["position_ids"] = torch.arange(0, batch['input_ids'].shape[1]).unsqueeze(0).to(self.model.device)
```
Comment (athitten) [inline nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:296]: @akoumpa your commits have removed the attention_mask so this line can potentially override the position_ids in case of packed sequence which is incorrect.

File: examples/llm/sft/automodel.py [hunk 7]
```diff
@@ -36,7 +36,14 @@
...
 def make_squad_hf_dataset(
-    tokenizer, micro_batch_size, seq_length, limit_dataset_samples=None, fp8=False, num_replicas=1, rank=0
+    tokenizer,
+    micro_batch_size,
+    seq_length,
+    packed_sequence_size,
+    limit_dataset_samples=None,
+    fp8=False,
+    num_replicas=1,
+    rank=0,
```
Comment (athitten) [inline examples/llm/sft/automodel.py:46]: Also previously the reason for getting a matching loss curve with PS and without was because of seq-len default being 2048 and also the packed sequence testing was done with packed-sequence-size 2048. So all the seqs in squad were getting padded upto 2048 leading to no packing at all and it ran similar to running with no packing. We should disable padding while running packed-sequences.

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py (hunk 6 above)
Comment (BoxiangW) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:296]: This line is needed for SP support actually. Also there is another CP-related change for `batch['position_ids']` as well. Will pack-seq input position_ids then? But this change is just for batch["position_ids"] to match the shape of `batch['input_ids']`. I think for pack-seq, we can address that instead

Comment (athitten) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:296]: Yes `PS` uses `position_ids` and the batch will contain `position_ids` while using `PS`. I was thinking just not do this line below if `"position_ids"` already exists in `batch` which is the case for `PS`. Does that sound okay ?
`batch["position_ids"] = torch.arange(0, batch['input_ids'].shape[1]).unsqueeze(0).to(self.model.device)`

Can you please share what code this is referring to ?
`Also there is another CP-related change for batch['position_ids'] as well.`

Comment (BoxiangW) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:296]: https://github.com/NVIDIA/NeMo/pull/12821/files#diff-a1007f10a48739b42901a65bd9baf0ccd0a4ae2f42bdf083eef8a684cba28360R280 This is the one for CP. Can you try run it with TP/SP and CP after ignoring them then? Loss curve would be affected if the shape does not match

Comment (athitten) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:296]: Do you mean run PS with TP/SP and CP ? If yes for PS we are just focusing on minimal functionality atm without any model parallelisms.

Comment (BoxiangW) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:297]: I see, thanks for the info

File: examples/llm/sft/automodel.py (hunk 7 above)
Comment (athitten) [reply examples/llm/sft/automodel.py:46]: Disabled padding if packed_sequence_size > 0: [here](https://github.com/NVIDIA/NeMo/pull/12551/commits/26bd344948111711d404364dea346c19194b8822#diff-6ca255a65c7e092b5f5720b3ab66bd9dd17cdc9d98ebea5f37789e1a48ec4c41R60)

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py (hunk 6 above)
Comment (athitten) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:297]: Added check to not override position_ids if it already exists in the batch: [here](https://github.com/NVIDIA/NeMo/pull/12551/commits/26bd344948111711d404364dea346c19194b8822#diff-a1007f10a48739b42901a65bd9baf0ccd0a4ae2f42bdf083eef8a684cba28360R296)

**Review by akoumpa — APPROVED** [review]

**Comment by chtruong814** [comment]
CI passed here except for an optional test.  Merging this now
https://github.com/NVIDIA/NeMo/actions/runs/14593384135/attempts/1
